### ISSUE TITLE
RxJs 5.5 lettable operators cause build error

### ISSUE THREAD

**Issue by deebloo** [angular/angular-cli#8165]
### Bug Report or Feature Request (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
```

### Versions.
```
Angular CLI: 1.5.0-rc.3
Node: 8.7.0
OS: darwin x64
Angular: 5.0.0-rc.3
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.5.0-rc.3
@angular-devkit/build-optimizer: 0.0.30
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.34
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0-rc.3
@schematics/angular: 0.0.48
typescript: 2.4.2
webpack: 3.8.1
```

### Repro steps.
1. Create new clean app
2. import and use any lettable rxjs operator

```TS
import { map } from 'rxjs/operators/map';

myObservable.pipe(map(item => item * 2));
```
3. ng serve / ng build

### The log given by the failure.
```
ERROR in ./src/app/app.component.ts
Module not found: Error: Can't resolve 'rxjs/operators/map' in '/Users/dannyblue/Documents/projects/rx-import-error/src/app'
resolve 'rxjs/operators/map' in '/Users/dannyblue/Documents/projects/rx-import-error/src/app'
  Parsed request is a module
  using description file: /Users/dannyblue/Documents/projects/rx-import-error/package.json (relative path: ./src/app)
    aliased with mapping 'rxjs/operators': '/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js' to '/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map'
      using description file: /Users/dannyblue/Documents/projects/rx-import-error/package.json (relative path: ./src/app)
        Field 'browser' doesn't contain a valid alias configuration
      after using description file: /Users/dannyblue/Documents/projects/rx-import-error/package.json (relative path: ./src/app)
        using description file: /Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/package.json (relative path: ./_esm5/operators/index.js/map)
          no extension
            Field 'browser' doesn't contain a valid alias configuration
            /Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map doesn't exist
          .ts
            Field 'browser' doesn't contain a valid alias configuration
            /Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map.ts doesn't exist
          .js
            Field 'browser' doesn't contain a valid alias configuration
            /Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map.js doesn't exist
          as directory
            /Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map doesn't exist
[/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map]
[/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map.ts]
[/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map.js]
[/Users/dannyblue/Documents/projects/rx-import-error/node_modules/rxjs/_esm5/operators/index.js/map]
 @ ./src/app/app.component.ts 14:12-41
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
```

### Mention any other details that might be useful.
[Link to repo](https://github.com/deebloo/rxjs-cli-error)

**Comment by ctrl-brk** [comment]
try to `import { map } from 'rxjs/operators';`

**Comment by deebloo** [comment]
@ctrl-brk If I do that the build works but the unused operators are still in the final bundle.
<img width="1440" alt="screen shot 2017-10-24 at 11 14 27 am" src="https://user-images.githubusercontent.com/6563621/31951818-15e00dd0-b8ad-11e7-9150-2a9f57686706.png">

**Comment by filipesilva** [comment]
Using the repo that @deebloo provided, I can see this behaviour.

```
git clone https://github.com/deebloo/rxjs-cli-error
cd rxjs-cli-error
npm i
npm run g -- build --prod --build-optimizer
```
```
$ npm run ng -- build --prod --build-optimizer

> rx-import-error@0.0.0 ng D:\sandbox\rxjs-cli-error
> ng "build" "--prod" "--build-optimizer"

Date: 2017-10-24T16:52:44.480Z
Hash: a4b7657f365a1661ca3b
Time: 24027ms
chunk {0} polyfills.656f60bebda0fce711d9.bundle.js (polyfills) 60.9 kB {3} [initial] [rendered]
chunk {1} main.af3c8bc916925fa41944.bundle.js (main) 925 bytes {3} [initial] [rendered]
chunk {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 0 bytes {3} [initial] [rendered]
chunk {3} inline.6a73046f6b90830eb75e.bundle.js (inline) 1.45 kB [entry] [rendered]

ERROR in ./src/app/app.component.ts
Module not found: Error: Can't resolve 'rxjs/operators/map' in 'D:\sandbox\rxjs-cli-error\src\app'
resolve 'rxjs/operators/map' in 'D:\sandbox\rxjs-cli-error\src\app'
  Parsed request is a module
  using description file: D:\sandbox\rxjs-cli-error\package.json (relative path: ./src/app)
    aliased with mapping 'rxjs/operators': 'D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js' to 'D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js/map'
      using description file: D:\sandbox\rxjs-cli-error\package.json (relative path: ./src/app)
        Field 'browser' doesn't contain a valid alias configuration
      after using description file: D:\sandbox\rxjs-cli-error\package.json (relative path: ./src/app)
        using description file: D:\sandbox\rxjs-cli-error\node_modules\rxjs\package.json (relative path: ./_esm5/operators/index.js/map)
          no extension
            Field 'browser' doesn't contain a valid alias configuration
            D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map doesn't exist
          .ts
            Field 'browser' doesn't contain a valid alias configuration
            D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map.ts doesn't exist
          .js
            Field 'browser' doesn't contain a valid alias configuration
            D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map.js doesn't exist
          as directory
            D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map doesn't exist
[D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map]
[D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map.ts]
[D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map.js]
[D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js\map]
 @ ./src/app/app.component.ts 5:0-41
 @ ./src/app/app.module.ngfactory.js
 @ ./src/main.ts
 @ multi ./src/main.ts
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! rx-import-error@0.0.0 ng: `ng "build" "--prod" "--build-optimizer"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the rx-import-error@0.0.0 ng script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\kamik\AppData\Roaming\npm-cache\_logs\2017-10-24T16_52_44_647Z-debug.log
```

Then I changed the import as shown in https://github.com/ReactiveX/rxjs/blob/master/doc/lettable-operators.md#build-and-treeshaking:
```
// ./src/app/app.component.ts
import { map } from 'rxjs/operators/map';
```
to
```
import { map } from 'rxjs/operators;
```
That fixes the error and gives me these bundle sizes:
```
Date: 2017-10-24T17:05:56.185Z
Hash: 194f0d7ab122d6f4f263
Time: 29615ms
chunk {0} polyfills.656f60bebda0fce711d9.bundle.js (polyfills) 60.9 kB {3} [initial] [rendered]
chunk {1} main.ac2c08dc874ec62b5060.bundle.js (main) 154 kB {3} [initial] [rendered]
chunk {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 0 bytes {3} [initial] [rendered]
chunk {3} inline.a5e27b649fb394f406e5.bundle.js (inline) 1.45 kB [entry] [rendered]
```

Then I tried using the full path again (`rxjs/operators/map`) and disabling the rxjs paths by editing the CLI build files directly. 
```
// ./node_modules/@angular/cli/models/webpack-configs/common.js
    try {
        // comment out these lines
        // const rxjsPathMappingImport = 'rxjs/_esm5/path-mapping';
        // const rxPaths = require_project_module_1.requireProjectModule(projectRoot, rxjsPathMappingImport);
        // alias = rxPaths(nodeModules);
    }
```

That gave me the following build size:
```
Date: 2017-10-24T17:16:13.485Z
Hash: 634e188023cbb2c12ba9
Time: 23337ms
chunk {0} polyfills.656f60bebda0fce711d9.bundle.js (polyfills) 60.9 kB {3} [initial] [rendered]
chunk {1} main.86b176fb160fb4a2b1b7.bundle.js (main) 166 kB {3} [initial] [rendered]
chunk {2} styles.d41d8cd98f00b204e980.bundle.css (styles) 0 bytes {3} [initial] [rendered]
chunk {3} inline.456823864452b4af648d.bundle.js (inline) 1.45 kB [entry] [rendered]
```

So I I think these are working as intended. Having the mapping makes it smaller. @deebloo can you have a look and see if there's something I'm missing in my testing procedure?

**Comment by devoto13** [comment]
It should work in both cases I believe. The problem seems quite clear from the error message @filipesilva posted.

    aliased with mapping 'rxjs/operators': 'D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js' to 'D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js/map'

In the `node_modules/_esm5/path-mapping.js` there is a dedicated entry for `rxjs/operators/map`, which looks like this: `"rxjs/operators/map": path.join(dir, "operators/map.js")`. However Webpack tries to apply less specific alias first (`"rxjs/operators": path.join(dir, "operators/index.js")`), which generates incorrect path - `D:\sandbox\rxjs-cli-error\node_modules\rxjs\_esm5\operators\index.js/map`.

And it's apparently reported here https://github.com/ReactiveX/rxjs/issues/2988.

**Comment by deebloo** [comment]
@filipesilva it looks like you are correct. I was doing tests without build optimizer. WITHOUT the build optimizer the values are.

Path Mapping:
<img width="1440" alt="screen shot 2017-10-24 at 1 47 28 pm" src="https://user-images.githubusercontent.com/6563621/31959205-0bccd0c0-b8c2-11e7-96af-18cbec9d3121.png">

NO Path Mapping
<img width="1440" alt="screen shot 2017-10-24 at 1 48 18 pm" src="https://user-images.githubusercontent.com/6563621/31959224-1aa23bc6-b8c2-11e7-9820-a5daf68bb86c.png">

**Comment by deebloo** [comment]
still having some size descrepencies in an actual app. still looking into it

**Comment by hansl** [comment]
This will be fixed in RXJS 5.5.2 (see https://github.com/ReactiveX/rxjs/pull/2987)

**Comment by hansl** [comment]
Duplicate of https://github.com/ReactiveX/rxjs/issues/2988
